src/main/kotlin/io/emeraldpay/dshackle/auth/processor/AuthProcessor.kt

```diff
@@ -0,0 +1,73 @@
+package io.emeraldpay.dshackle.auth.processor
+
+import com.auth0.jwt.JWT
+import com.auth0.jwt.JWTVerifier
+import com.auth0.jwt.algorithms.Algorithm
+import io.emeraldpay.dshackle.auth.AuthContext
+import io.emeraldpay.dshackle.auth.service.KeyReader
+import io.grpc.Status
+import org.springframework.stereotype.Component
+import java.security.PrivateKey
+import java.security.PublicKey
+import java.security.interfaces.RSAPrivateKey
+import java.security.interfaces.RSAPublicKey
+import java.time.Instant
+import java.util.UUID
+
+const val SESSION_ID = "sessionId"
+const val VERSION = "version"
+
+enum class AuthVersion {
+    V1;
+
+    companion object {
+        fun getVersion(version: String) = values().find { it.name == version }
+            ?: throw Status.INVALID_ARGUMENT
+                .withDescription("Unsupported auth version $version")
+                .asException()
+    }
+}
+
+abstract class AuthProcessor {
+
+    open fun process(keyPair: KeyReader.KeyPair, token: String): String {
+        try {
+            val verifier: JWTVerifier = JWT.require(verifyingAlgorithm(keyPair.drpcPublicKey))
+                .withIssuer("drpc")
+                .build()
+            verifier.verify(token)
+        } catch (e: Exception) {
+            throw Status.INVALID_ARGUMENT
+                .withDescription("Invalid token: ${e.message}")
+                .asException()
+        }
+
+        return processInternal(keyPair.providerPrivateKey)
+            .also {
+                AuthContext.putTokenInContext(it)
```

Comment:
It's very strange place to put something in context, we should do this in AuthService may be?

Reply:
Agree, fixed